cmd/query/app/server.go

```diff
@@ -316,9 +343,15 @@ func (s *Server) Start() error {
 				s.logger.Error("Could not start multiplexed server", zap.Error(err))
 			}
 			s.unavailableChannel <- healthcheck.Unavailable
+			wg.Done()
 		}()
 	}
 
+	go func() {
+		wg.Wait()
+		close(s.unavailableChannel)
```

Comment:
I don't think we could do this in Close() because the various server goroutines (http, grpc, cmux) are what get shutdown by the call to Close(), at which point they all try and _write_ to `s.unavailableChannel`, so if this has already been closed, it would cause a panic. That's why I think this channel should only be closed _as a consequence of the server goroutines returning_ rather than as a direct consequence of Close().

Just in case you missed it I described the shutdown flow in more detail in the commit message of https://github.com/jaegertracing/jaeger/pull/5299/commits/faf5127498c402eb97feec5226201284f8155315.

There might be a different architecture we could switch to that would simplify things, but this is the best I could come up with while trying to keep the code changes minimal.

Reply:
I am not suggesting closing the channel first, the order could be
  * close http server
  * close grpc server
  * wait for Done
  * close error channel

Overall though, I do agree that this error channel design is flimsy, easy to deadlock if nobody is reading it. IIRC in the OTEL collector they just pass a callback function to components to report their status and the implementor of that callback can run its own channel if necessary. I find channels as an API mechanism pretty inconvenient to work with.

Reply:
Thanks for the suggestion. I've had a go at doing this in https://github.com/jaegertracing/jaeger/pull/5299/commits/494415ef8ed1a59ccc31b68d015dda323c8ab227, which I think significantly simplifies things, but also means the user of the API needs to be more careful because they _must_ ensure the `unavailableChannel` if they're going to call `Stop` (or it'll block forever).

Definitely agree with you more broadly about the clunkiness of the current API and a callback approach would probably be nicer and more flexible. However I also think it would be nice to leave any more substantial refactor for future work to avoid scope creep in this PR.

Reply:
Addressed by the merge of main of changes in https://github.com/jaegertracing/jaeger/issues/5307.